<a href="https://colab.research.google.com/github/nchibana/DS-Unit-1-Sprint-3-Statistical-Tests-and-Experiments/blob/master/module1-statistics-probability-and-inference/LS_DS_131_Statistics_Probability_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 3 Assignment 1*

# Apply the t-test to real data

Your assignment is to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

Your goals:

1. Load and clean the data (or determine the best method to drop observations when running tests)
2. Using hypothesis testing, find an issue that democrats support more than republicans with p < 0.01
3. Using hypothesis testing, find an issue that republicans support more than democrats with p < 0.01
4. Using hypothesis testing, find an issue where the difference between republicans and democrats has p > 0.1 (i.e. there may not be much of a difference)

Note that this data will involve *2 sample* t-tests, because you're comparing averages across two groups (republicans and democrats) rather than a single group against a null hypothesis.

Stretch goals:

1. Refactor your code into functions so it's easy to rerun with arbitrary variables
2. Apply hypothesis testing to your personal project data (for the purposes of this notebook you can type a summary of the hypothesis you formed and tested)

In [0]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv("https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data", header=None)
df.head()

0  1  2  3  4  5  6  7  8  9  10 11 12 13 14 15 16
0  republican  n  y  n  y  y  y  n  n  n  y  ?  y  y  y  n  y
1  republican  n  y  n  y  y  y  n  n  n  n  n  y  y  y  n  ?
2    democrat  ?  y  y  ?  y  y  n  n  n  n  y  n  y  y  n  n
3    democrat  n  y  y  n  ?  y  n  n  n  n  y  n  y  n  n  y
4    democrat  y  y  y  n  y  y  n  n  n  n  y  ?  y  y  y  y

In [3]:
df = df.rename(columns={0: 'party', 
                        1: 'handicapped-infants', 
                        2: 'water-project', 
                        3: 'budget', 
                        4: 'physician-fee-freeze', 
                        5:'el-salvador-aid', 
                        6: 'religious-groups', 
                        7: 'anti-satellite-test-ban', 
                        8: 'aid-to-contras', 
                        9: 'mx-missile', 
                        10: 'immigration', 
                        11: 'synfuels', 
                        12: 'education', 
                        13: 'right-to-sue', 
                        14:'crime', 
                        15: 'duty-free', 
                        16: 'south-africa'})

df.head()


party handicapped-infants water-project  ... crime duty-free south-africa
0  republican                   n             y  ...     y         n            y
1  republican                   n             y  ...     y         n            ?
2    democrat                   ?             y  ...     y         n            n
3    democrat                   n             y  ...     n         n            y
4    democrat                   y             y  ...     y         y            y

[5 rows x 17 columns]

In [4]:
df = df.replace(['n', 'y'], [0, 1])
print(df.shape)
df.head()

(435, 17)


party handicapped-infants water-project  ... crime duty-free south-africa
0  republican                   0             1  ...     1         0            1
1  republican                   0             1  ...     1         0            ?
2    democrat                   ?             1  ...     1         0            0
3    democrat                   0             1  ...     0         0            1
4    democrat                   1             1  ...     1         1            1

[5 rows x 17 columns]

In [10]:
df = df.replace('?', np.NaN)
df.head()

party  handicapped-infants  ...  duty-free  south-africa
0  republican                  0.0  ...        0.0           1.0
1  republican                  0.0  ...        0.0           NaN
2    democrat                  NaN  ...        0.0           0.0
3    democrat                  0.0  ...        0.0           1.0
4    democrat                  1.0  ...        1.0           1.0

[5 rows x 17 columns]

In [12]:
df.isna().sum()

party                        0
handicapped-infants         12
water-project               48
budget                      11
physician-fee-freeze        11
el-salvador-aid             15
religious-groups            11
anti-satellite-test-ban     14
aid-to-contras              15
mx-missile                  22
immigration                  7
synfuels                    21
education                   31
right-to-sue                25
crime                       17
duty-free                   28
south-africa               104
dtype: int64

In [0]:
rep = df[df['party'] == 'republican']
dem = df[df['party'] == 'democrat']

In [0]:
from scipy.stats import ttest_ind, ttest_ind_from_stats, ttest_rel

##MX-Missile Bill##

In [22]:
print("Democrat Support: ", dem['mx-missile'].mean())
print("Republican Support: ", rep['mx-missile'].mean())

Democrat Support:  0.7580645161290323
Republican Support:  0.11515151515151516


In [25]:
ttest_ind(dem['mx-missile'], rep['mx-missile'], nan_policy='omit')

Ttest_indResult(statistic=16.437503268542994, pvalue=5.03079265310811e-47)

-Null hypothesis: The mean of Democrat support is equal to the mean of Republican support for this bill.

-Alternative hypothesis: The mean of Democrat support is not the same as the mean of Republican support for this bill.

-T-statistic: 16.4375

-P-value: 5.03e-47 (nearly 0)

-Confidence level: 1 - 5.03e-47 (nearly 1)

**Given the results of the above test, I would REJECT the null hypothesis that there is the same level of support for this bill among Democrats and Republicans**

##El Salvador Bill##

In [20]:
print("Democrat Support: ", dem['el-salvador-aid'].mean())
print("Republican Support: ", rep['el-salvador-aid'].mean())

Democrat Support:  0.21568627450980393
Republican Support:  0.9515151515151515


In [21]:
ttest_ind(dem['el-salvador-aid'], rep['el-salvador-aid'], nan_policy="omit")

Ttest_indResult(statistic=-21.13669261173219, pvalue=5.600520111729011e-68)

-Null hypothesis: The mean of Democrat support is equal to the mean of Republican support for this bill.

-Alternative hypothesis: The mean of Democrat support is not the same as the mean of Republican support for this bill.

-T-statistic: -21.13669

-P-value: 5.6e-68 (nearly 0)

-Confidence level: 1 -  5.6e-68 (nearly 1)

**Given the results of the above test, I would REJECT the null hypothesis that there is the same level of support for this bill among Democrats and Republicans**

##Water Project BIll

In [68]:
print("Democrat Support: ", dem['water-project'].mean())
print("Republican Support: ", rep['water-project'].mean())

Democrat Support:  0.502092050209205
Republican Support:  0.5067567567567568


In [69]:
ttest_ind(dem['water-project'], rep['water-project'], nan_policy="omit")

Ttest_indResult(statistic=-0.08896538137868286, pvalue=0.9291556823993485)

-Null hypothesis: The mean of Democrat support is equal to the mean of Republican support for this bill.

-Alternative hypothesis: The mean of Democrat support is not the same as the mean of Republican support for this bill.

-T-statistic: -0.088965

-P-value: 0.92915568

-Confidence level: 1 - 0.9291556

**Given the results of the above test, I would FAIL TO REJECT the null hypothesis that there is the same level of support for this bill among Democrats and Republicans**

In [0]:
def t_test(x):
  results = ttest_ind(dem[x], rep[x], nan_policy="omit")
  p_value = results.pvalue
  
  if p_value < 0.01:
    print("Given the results of the above test, I would REJECT the null hypothesis that there is the same level of support for this bill among Democrats and Republicans")
  else:
    print("Given the results of the above test, I would FAIL TO REJECT the null hypothesis that there is the same level of support for this bill among Democrats and Republicans")
  

In [102]:
t_test('water-project')

Given the results of the above test, I would FAIL TO REJECT the null hypothesis that there is the same level of support for this bill among Democrats and Republicans


In [103]:
t_test('el-salvador-aid')

Given the results of the above test, I would REJECT the null hypothesis that there is the same level of support for this bill among Democrats and Republicans


In [104]:
t_test('handicapped-infants')

Given the results of the above test, I would REJECT the null hypothesis that there is the same level of support for this bill among Democrats and Republicans


In [86]:
t_test('budget')

Given the results of the above test, I would REJECT the null hypothesis that there is the same level of support for this bill among Democrats and Republicans


In [87]:
t_test('physician-fee-freeze')

Given the results of the above test, I would REJECT the null hypothesis that there is the same level of support for this bill among Democrats and Republicans


In [88]:
t_test('religious-groups')

Given the results of the above test, I would REJECT the null hypothesis that there is the same level of support for this bill among Democrats and Republicans


In [89]:
t_test('anti-satellite-test-ban')

Given the results of the above test, I would REJECT the null hypothesis that there is the same level of support for this bill among Democrats and Republicans


In [90]:
t_test('aid-to-contras')

Given the results of the above test, I would REJECT the null hypothesis that there is the same level of support for this bill among Democrats and Republicans


In [91]:
t_test('mx-missile')

Given the results of the above test, I would REJECT the null hypothesis that there is the same level of support for this bill among Democrats and Republicans


In [105]:
t_test('immigration')

Given the results of the above test, I would FAIL TO REJECT the null hypothesis that there is the same level of support for this bill among Democrats and Republicans


In [93]:
t_test('synfuels')

Given the results of the above test, I would REJECT the null hypothesis that there is the same level of support for this bill among Democrats and Republicans


In [95]:
t_test('education')

Given the results of the above test, I would REJECT the null hypothesis that there is the same level of support for this bill among Democrats and Republicans


In [96]:
t_test('right-to-sue')

Given the results of the above test, I would REJECT the null hypothesis that there is the same level of support for this bill among Democrats and Republicans


In [97]:
t_test('crime')

Given the results of the above test, I would REJECT the null hypothesis that there is the same level of support for this bill among Democrats and Republicans


In [98]:
t_test('duty-free')

Given the results of the above test, I would REJECT the null hypothesis that there is the same level of support for this bill among Democrats and Republicans


In [100]:
t_test('south-africa')

Given the results of the above test, I would REJECT the null hypothesis that there is the same level of support for this bill among Democrats and Republicans
